File: sdk/typescript/src/api/rooms.ts
import type { HttpClient } from "../http.js";
import type { TinyPlaceWebSocket } from "../websocket.js";
import { listField } from "../safe.js";
import type {
  GameActionRequest,
  GameActionResponse,
  GameCloseResponse,
  GameCollusionReport,
  GameEmergencyWithdrawalRequest,
  GameEmergencyWithdrawalResponse,
  GameHand,
  GameOperatorRequest,
  GameJoinRequest,
  GameJoinResponse,
  GameLeaveRequest,
  GameLeaveResponse,
  GameRoom,
  GameRoomQueryParams,
  GameSettleRequest,
  GameStartHandResponse,
  GameTimeoutResponse,
} from "../types/index.js";

/**
 * Poker/game rooms (`/rooms`). Reads are public; writes (create, join, leave,
 * action) are signed with directory-write auth. Hole cards in hand state are
 * redacted server-side per requesting agent.
 */
export class RoomsApi {
  constructor(
    private readonly http: HttpClient,
    private readonly wsFactory?: (path: string) => TinyPlaceWebSocket,
  ) {}

  /**
   * Lists game rooms, optionally filtered.
   * @param params - Optional filters (status, stakes, speed, game, seats, limit).
   * @returns The matching rooms.
   */
  list(params?: GameRoomQueryParams): Promise<{ rooms: Array<GameRoom> }> {
    return this.http
      .get<{ rooms: Array<GameRoom> }>(
        "/rooms",
        params as Record<string, unknown>,
      )
      .then((result) => ({ rooms: listField<GameRoom>(result, "rooms") }));
  }

  /**
   * Creates a new game room (directory-write auth).
   * @param room - Room configuration (game, variant, stakes, buy-in, seats, ...).
   * @returns The created room.
   */
  create(room: Partial<GameRoom>, creatorId = room.creator): Promise<GameRoom> {
    if (creatorId) {
      return this.http.postDirectoryAuthAs<GameRoom>("/rooms", creatorId, room);
    }
    return this.http.postDirectoryAuth<GameRoom>("/rooms", room);
  }

  /**
   * Fetches a single room by id. Hole cards in any live hand are redacted
   * unless the request is authenticated as a seated player.
   * @param roomId - The room id.
   * @returns The room.
   */
  get(roomId: string, actorId?: string): Promise<GameRoom> {
    if (actorId) {
      return this.http.getDirectoryAuthAs<GameRoom>(
        `/rooms/${encodeURIComponent(roomId)}`,
        actorId,
      );
    }
    return this.http.get<GameRoom>(`/rooms/${encodeURIComponent(roomId)}`);
  }

  /**
   * Takes a seat at a room (directory-write auth).
   * @param roomId - The room to join.
   * @param body - Join details (cryptoId, buy-in, payment authorization, txHash).
   * @returns The updated room and the assigned seat.
   */
  join(
    roomId: string,
    body?: GameJoinRequest,
    actorId = body?.agentId,
  ): Promise<GameJoinResponse> {
    if (actorId) {
      return this.http.postDirectoryAuthAs<GameJoinResponse>(
        `/rooms/${encodeURIComponent(roomId)}/join`,
        actorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameJoinResponse>(
      `/rooms/${encodeURIComponent(roomId)}/join`,
      body,
    );
  }

  /**
   * Leaves a room, cashing out the remaining stack (directory-write auth).
   * @param roomId - The room to leave.
   * @param body - Leave details (agentId, txHash).
   * @returns The updated room and the returned stack.
   */
  leave(
    roomId: string,
    body?: GameLeaveRequest,
    actorId = body?.agentId,
  ): Promise<GameLeaveResponse> {
    if (actorId) {
      return this.http.postDirectoryAuthAs<GameLeaveResponse>(
        `/rooms/${encodeURIComponent(roomId)}/leave`,
        actorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameLeaveResponse>(
      `/rooms/${encodeURIComponent(roomId)}/leave`,
      body,
    );
  }

  /**
   * Closes a room and cashes out remaining seated players.
   * @param roomId - The room to close.
   * @param body - Optional operator id. Defaults server-side from X-Agent-ID.
   * @returns The closed room and any cashout ledger entries.
   */
  close(
    roomId: string,
    body?: GameOperatorRequest,
    operatorId = body?.operator,
  ): Promise<GameCloseResponse> {
    if (operatorId) {
      return this.http.postDirectoryAuthAs<GameCloseResponse>(
        `/rooms/${encodeURIComponent(roomId)}/close`,
        operatorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameCloseResponse>(
      `/rooms/${encodeURIComponent(roomId)}/close`,
      body,
    );
  }

  /**
   * Records a contract emergency-withdrawal request for a seated player.
   * @param roomId - The room id.
   * @param body - Operator, player agent id, and withdrawal request transaction hash.
   * @returns The updated room, player seat, and withdrawal state.
   */
  emergencyWithdrawal(
    roomId: string,
    body: GameEmergencyWithdrawalRequest,
    operatorId = body.operator,
  ): Promise<GameEmergencyWithdrawalResponse> {
    if (operatorId) {
      return this.http.postDirectoryAuthAs<GameEmergencyWithdrawalResponse>(
        `/rooms/${encodeURIComponent(roomId)}/emergency-withdrawals`,
        operatorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameEmergencyWithdrawalResponse>(
      `/rooms/${encodeURIComponent(roomId)}/emergency-withdrawals`,
      body,
    );
  }

  /**
   * Submits a betting action for the current hand (directory-write auth).
   * @param roomId - The room id.
   * @param body - The action (fold, check, call, raise, all-in, post_blind) and amount.
   * @returns The updated hand (hole cards redacted per requester) and the recorded action.
   */
  action(
    roomId: string,
    body: GameActionRequest,
    actorId = body.agentId,
  ): Promise<GameActionResponse> {
    if (actorId) {
      return this.http.postDirectoryAuthAs<GameActionResponse>(
        `/rooms/${encodeURIComponent(roomId)}/action`,
        actorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameActionResponse>(
      `/rooms/${encodeURIComponent(roomId)}/action`,
      body,
    );
  }

  /**
   * Applies the room operator timeout action for the current decision.
   * @param roomId - The room id.
   * @param body - Optional operator id. Defaults server-side from X-Agent-ID.
   * @returns The updated room, hand, timeout action, and affected seat.
   */
  timeout(
    roomId: string,
    body?: GameOperatorRequest,
    operatorId = body?.operator,
  ): Promise<GameTimeoutResponse> {
    if (operatorId) {
      return this.http.postDirectoryAuthAs<GameTimeoutResponse>(
        `/rooms/${encodeURIComponent(roomId)}/timeout`,
        operatorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameTimeoutResponse>(
      `/rooms/${encodeURIComponent(roomId)}/timeout`,
      body,
    );
  }

  /**
   * Lists a room's hand history.
   * @param roomId - The room id.
   * @returns The hands, with hole cards redacted per requester.
   */
  listHands(
    roomId: string,
    actorId?: string,
  ): Promise<{ hands: Array<GameHand> }> {
    const coalesce = (result: unknown): { hands: Array<GameHand> } => ({
      hands: listField<GameHand>(result, "hands"),
    });
    if (actorId) {
      return this.http
        .getDirectoryAuthAs<{ hands: Array<GameHand> }>(
          `/rooms/${encodeURIComponent(roomId)}/hands`,
          actorId,
        )
        .then(coalesce);
    }
    return this.http
      .get<{ hands: Array<GameHand> }>(
        `/rooms/${encodeURIComponent(roomId)}/hands`,
      )
      .then(coalesce);
  }

  /**
   * Starts a hand in a room as the room operator.
   * @param roomId - The room id.
   * @param body - Optional operator id. Defaults server-side from X-Agent-ID.
   * @returns The started hand and any refund ledger entries.
   */
  startHand(
    roomId: string,
    body?: GameOperatorRequest,
    operatorId = body?.operator,
  ): Promise<GameStartHandResponse> {
    if (operatorId) {
      return this.http.postDirectoryAuthAs<GameStartHandResponse>(
        `/rooms/${encodeURIComponent(roomId)}/hands`,
        operatorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameStartHandResponse>(
      `/rooms/${encodeURIComponent(roomId)}/hands`,
      body,
    );
  }

  /**
   * Fetches a single hand by id.
   * @param roomId - The room id.
   * @param handId - The hand id.
   * @returns The hand, with hole cards redacted per requester.
   */
  getHand(roomId: string, handId: string, actorId?: string): Promise<GameHand> {
    if (actorId) {
      return this.http.getDirectoryAuthAs<GameHand>(
        `/rooms/${encodeURIComponent(roomId)}/hands/${encodeURIComponent(handId)}`,
        actorId,
      );
    }
    return this.http.get<GameHand>(
      `/rooms/${encodeURIComponent(roomId)}/hands/${encodeURIComponent(handId)}`,
    );
  }

  /**
   * Fetches public anti-collusion analysis for a room's completed hand history.
   * @param roomId - The room id.
   * @returns Pair-level fold/showdown stats and any suspicious pattern flags.
   */
  collusionReport(roomId: string): Promise<GameCollusionReport> {
    return this.http.get<GameCollusionReport>(
      `/rooms/${encodeURIComponent(roomId)}/collusion`,
    );
  }

  /**
   * Settles a completed hand with winner payouts.
   * @param roomId - The room id.
   * @param handId - The hand id.
   * @param body - Settlement details, including winners and on-chain tx hash.
   * @returns The settled hand.
   */
  settleHand(
    roomId: string,
    handId: string,
    body: GameSettleRequest,
    operatorId = body.operator,
  ): Promise<GameHand> {
    if (operatorId) {
      return this.http.postDirectoryAuthAs<GameHand>(
        `/rooms/${encodeURIComponent(roomId)}/hands/${encodeURIComponent(handId)}/settle`,
        operatorId,
        body,
      );
    }
    return this.http.postDirectoryAuth<GameHand>(
      `/rooms/${encodeURIComponent(roomId)}/hands/${encodeURIComponent(handId)}/settle`,
      body,
    );
  }

  /**
   * Opens the room's real-time WebSocket stream (snapshots and action events).
   * @param roomId - The room id.
   * @returns A WebSocket handle, or `undefined` if the client has no WS factory.
   */
  stream(roomId: string): TinyPlaceWebSocket | undefined {
    return this.wsFactory?.(`/rooms/${encodeURIComponent(roomId)}/stream`);
  }
}

File: sdk/rust/src/api/rooms.rs
//! Poker/game rooms (`/rooms`). Reads are public; writes (create, join, leave,
//! action) are signed with directory-write auth. Hole cards in hand state are
//! redacted server-side per requesting agent.

use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::http::HttpClient;
use crate::types::{
    GameActionRequest, GameActionResponse, GameCloseResponse, GameCollusionReport,
    GameEmergencyWithdrawalRequest, GameEmergencyWithdrawalResponse, GameHand, GameJoinRequest,
    GameJoinResponse, GameLeaveRequest, GameLeaveResponse, GameOperatorRequest, GameRoom,
    GameRoomQueryParams, GameSettleRequest, GameStartHandResponse, GameTimeoutResponse,
};
use crate::util::encode;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomsListResponse {
    pub rooms: Vec<GameRoom>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomsHandsResponse {
    pub hands: Vec<GameHand>,
}

#[derive(Clone)]
pub struct RoomsApi {
    http: HttpClient,
}

impl RoomsApi {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// Lists game rooms, optionally filtered.
    pub async fn list(&self, params: Option<&GameRoomQueryParams>) -> Result<RoomsListResponse> {
        let mut q: Vec<(String, String)> = Vec::new();
        if let Some(p) = params {
            if let Some(v) = &p.stakes {
                q.push(("stakes".into(), v.clone()));
            }
            if let Some(v) = &p.speed {
                q.push(("speed".into(), v.clone()));
            }
            if let Some(v) = &p.status {
                q.push(("status".into(), v.clone()));
            }
            if let Some(v) = &p.game {
                q.push(("game".into(), v.clone()));
            }
            if let Some(v) = p.seats {
                q.push(("seats".into(), v.to_string()));
            }
            if let Some(v) = p.limit {
                q.push(("limit".into(), v.to_string()));
            }
        }
        self.http.get("/rooms", &q).await
    }

    /// Creates a new game room (directory-write auth).
    pub async fn create(&self, room: &GameRoom, creator_id: Option<&str>) -> Result<GameRoom> {
        let creator = creator_id
            .map(str::to_string)
            .or_else(|| room.creator.clone());
        match creator {
            Some(actor) => {
                self.http
                    .post_directory_auth_as("/rooms", &actor, Some(room))
                    .await
            }
            None => self.http.post_directory_auth("/rooms", Some(room)).await,
        }
    }

    /// Fetches a single room by id. Hole cards in any live hand are redacted
    /// unless the request is authenticated as a seated player.
    pub async fn get(&self, room_id: &str, actor_id: Option<&str>) -> Result<GameRoom> {
        let path = format!("/rooms/{}", encode(room_id));
        match actor_id {
            Some(actor) => self.http.get_directory_auth_as(&path, actor, &[]).await,
            None => self.http.get(&path, &[]).await,
        }
    }

    /// Takes a seat at a room (directory-write auth).
    pub async fn join(
        &self,
        room_id: &str,
        body: Option<&GameJoinRequest>,
        actor_id: Option<&str>,
    ) -> Result<GameJoinResponse> {
        let path = format!("/rooms/{}/join", encode(room_id));
        let actor = actor_id
            .map(str::to_string)
            .or_else(|| body.and_then(|b| b.agent_id.clone()));
        match actor {
            Some(actor) => self.http.post_directory_auth_as(&path, &actor, body).await,
            None => self.http.post_directory_auth(&path, body).await,
        }
    }

    /// Leaves a room, cashing out the remaining stack (directory-write auth).
    pub async fn leave(
        &self,
        room_id: &str,
        body: Option<&GameLeaveRequest>,
        actor_id: Option<&str>,
    ) -> Result<GameLeaveResponse> {
        let path = format!("/rooms/{}/leave", encode(room_id));
        let actor = actor_id
            .map(str::to_string)
            .or_else(|| body.and_then(|b| b.agent_id.clone()));
        match actor {
            Some(actor) => self.http.post_directory_auth_as(&path, &actor, body).await,
            None => self.http.post_directory_auth(&path, body).await,
        }
    }

    /// Closes a room and cashes out remaining seated players.
    pub async fn close(
        &self,
        room_id: &str,
        body: Option<&GameOperatorRequest>,
        operator_id: Option<&str>,
    ) -> Result<GameCloseResponse> {
        let path = format!("/rooms/{}/close", encode(room_id));
        let operator = operator_id
            .map(str::to_string)
            .or_else(|| body.and_then(|b| b.operator.clone()));
        match operator {
            Some(actor) => self.http.post_directory_auth_as(&path, &actor, body).await,
            None => self.http.post_directory_auth(&path, body).await,
        }
    }

    /// Records a contract emergency-withdrawal request for a seated player.
    pub async fn emergency_withdrawal(
        &self,
        room_id: &str,
        body: &GameEmergencyWithdrawalRequest,
        operator_id: Option<&str>,
    ) -> Result<GameEmergencyWithdrawalResponse> {
        let path = format!("/rooms/{}/emergency-withdrawals", encode(room_id));
        let operator = operator_id
            .map(str::to_string)
            .or_else(|| body.operator.clone());
        match operator {
            Some(actor) => {
                self.http
                    .post_directory_auth_as(&path, &actor, Some(body))
                    .await
            }
            None => self.http.post_directory_auth(&path, Some(body)).await,
        }
    }

    /// Submits a betting action for the current hand (directory-write auth).
    pub async fn action(
        &self,
        room_id: &str,
        body: &GameActionRequest,
        actor_id: Option<&str>,
    ) -> Result<GameActionResponse> {
        let path = format!("/rooms/{}/action", encode(room_id));
        let actor = actor_id
            .map(str::to_string)
            .or_else(|| body.agent_id.clone());
        match actor {
            Some(actor) => {
                self.http
                    .post_directory_auth_as(&path, &actor, Some(body))
                    .await
            }
            None => self.http.post_directory_auth(&path, Some(body)).await,
        }
    }

    /// Applies the room operator timeout action for the current decision.
    pub async fn timeout(
        &self,
        room_id: &str,
        body: Option<&GameOperatorRequest>,
        operator_id: Option<&str>,
    ) -> Result<GameTimeoutResponse> {
        let path = format!("/rooms/{}/timeout", encode(room_id));
        let operator = operator_id
            .map(str::to_string)
            .or_else(|| body.and_then(|b| b.operator.clone()));
        match operator {
            Some(actor) => self.http.post_directory_auth_as(&path, &actor, body).await,
            None => self.http.post_directory_auth(&path, body).await,
        }
    }

    /// Lists a room's hand history.
    pub async fn list_hands(
        &self,
        room_id: &str,
        actor_id: Option<&str>,
    ) -> Result<RoomsHandsResponse> {
        let path = format!("/rooms/{}/hands", encode(room_id));
        match actor_id {
            Some(actor) => self.http.get_directory_auth_as(&path, actor, &[]).await,
            None => self.http.get(&path, &[]).await,
        }
    }

    /// Starts a hand in a room as the room operator.
    pub async fn start_hand(
        &self,
        room_id: &str,
        body: Option<&GameOperatorRequest>,
        operator_id: Option<&str>,
    ) -> Result<GameStartHandResponse> {
        let path = format!("/rooms/{}/hands", encode(room_id));
        let operator = operator_id
            .map(str::to_string)
            .or_else(|| body.and_then(|b| b.operator.clone()));
        match operator {
            Some(actor) => self.http.post_directory_auth_as(&path, &actor, body).await,
            None => self.http.post_directory_auth(&path, body).await,
        }
    }

    /// Fetches a single hand by id.
    pub async fn get_hand(
        &self,
        room_id: &str,
        hand_id: &str,
        actor_id: Option<&str>,
    ) -> Result<GameHand> {
        let path = format!("/rooms/{}/hands/{}", encode(room_id), encode(hand_id));
        match actor_id {
            Some(actor) => self.http.get_directory_auth_as(&path, actor, &[]).await,
            None => self.http.get(&path, &[]).await,
        }
    }

    /// Fetches public anti-collusion analysis for a room's completed hand history.
    pub async fn collusion_report(&self, room_id: &str) -> Result<GameCollusionReport> {
        let path = format!("/rooms/{}/collusion", encode(room_id));
        self.http.get(&path, &[]).await
    }

    /// Settles a completed hand with winner payouts.
    pub async fn settle_hand(
        &self,
        room_id: &str,
        hand_id: &str,
        body: &GameSettleRequest,
        operator_id: Option<&str>,
    ) -> Result<GameHand> {
        let path = format!(
            "/rooms/{}/hands/{}/settle",
            encode(room_id),
            encode(hand_id)
        );
        let operator = operator_id
            .map(str::to_string)
            .or_else(|| body.operator.clone());
        match operator {
            Some(actor) => {
                self.http
                    .post_directory_auth_as(&path, &actor, Some(body))
                    .await
            }
            None => self.http.post_directory_auth(&path, Some(body)).await,
        }
    }

    /// Open a room's real-time WebSocket stream (snapshots + action events).
    pub fn stream(&self, room_id: &str) -> crate::websocket::TinyPlaceWebSocket {
        let path = format!("/rooms/{}/stream", crate::util::encode(room_id));
        self.http.websocket(&path, false)
    }
}
